.github/codeql/queries/token-not-used.ql

```diff
@@ -0,0 +1,41 @@
+/**
+ * @name Don't ignore the token for a cancelable progress bar
+ * @kind problem
+ * @problem.severity warning
+ * @id vscode-codeql/token-not-used
+ * @description If we call `withProgress` with `cancellable: true` but then
+ * ignore the token that is given to us, it will lead to a poor user experience
+ * because the progress bar will appear to be canceled but it will not actually
+ * affect the background process.
+ */
+
+import javascript
+
+class NewTokenSource extends CallExpr {
+  NewTokenSource() {
+    this.getCalleeName() = "withProgress" or this.getCalleeName() = "withInheritedProgress"
+  }
+
+  Function getCallback() {
+    this.getCalleeName() = "withProgress" and result = this.getArgument(0)
+    or
+    this.getCalleeName() = "withInheritedProgress" and result = this.getArgument(1)
+  }
+
+  ObjectExpr getOptions() {
+    this.getCalleeName() = "withProgress" and result = this.getArgument(1)
+    or
+    this.getCalleeName() = "withInheritedProgress" and result = this.getArgument(2)
+  }
+
+  predicate usesToken() { this.getCallback().getNumParameter() >= 2 }
+
+  predicate isCancellable() {
+    this.getOptions().getPropertyByName("cancellable").getInit().(BooleanLiteral).getBoolValue() =
+      true
+  }
+}
+
+from NewTokenSource t
+where t.isCancellable() and not t.usesToken()
+select t, "This progress bar is cancelable but the token is not used"
```

Comment:
I think it might be nice to make `cancellable` an entity here so it's clickable. Something like this:

```suggestion
select t, "This progress bar is $@ but the token is not used", t.getCancellableProperty(),
  t.getCancellableProperty().getName()
```

I'll leave a comment on `isCancellable` for how I think `getCancellableProperty` can be implemented.

Reply:
I've ended up only putting this link into the query for when the progress bar is cancellable because then there's always something to link to. For the other query there usually won't be because it's rare to explicitly do `cancellable: false`.